genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,554 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Package genwinsyscallproto generates Windows syscall function prototypes ("//sys ..." comments)
+// using given win32metadata information parsed by go-winmd as specified by ECMA-335.
+package genwinsyscallproto
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"go/token"
+	"io"
+	"log"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/microsoft/go-winmd"
+	"github.com/microsoft/go-winmd/coded"
+	"github.com/microsoft/go-winmd/flags"
+)
+
+// Context stores data about syscall generation so far, and to improve generation performance. It
+// keeps track of the list of used typedefs that may need to also be defined in generated Go code.
+// For performance, indexing the data helps with (e.g.) traversing one-way pointers backwards rather
+// than scanning the entire table each time.
+type Context struct {
+	Metadata *winmd.Metadata
+
+	// UsedTypeRefs	is the set of all TypeRefs used by syscall function parameters. Go
+	// definitions of these refs also need to be generated to make the syscalls usable.
+	UsedTypeRefs map[winmd.Index]struct{}
+
+	// AllTypeDefs is the set of all TypeDefs in the winmd file indexed by namespace + "::" + name.
+	AllTypeDefs map[string]winmd.Index
+	// MethodDefImplMap maps MethodDef -> ImplMap with matching MemberForwarded index.
+	MethodDefImplMap map[winmd.Index]*winmd.ImplMap
+	// FieldConstant maps Field -> Constant with the field as its parent.
+	FieldConstant map[winmd.Index]*winmd.Constant
+	// TypeDefNativeTypedefAttribute maps TypeDef -> CustomAttribute (if NativeTypedefAttribute).
+	TypeDefNativeTypedefAttribute map[winmd.Index]*winmd.CustomAttribute
+}
+
+// NewContext creates a Context. Performs some pre-processing to improve generation performance.
+func NewContext(f *winmd.Metadata) (*Context, error) {
+	l := &Context{
+		Metadata: f,
+
+		UsedTypeRefs: make(map[winmd.Index]struct{}),
+
+		AllTypeDefs:                   make(map[string]winmd.Index, f.Tables.TypeDef.Len),
+		MethodDefImplMap:              make(map[winmd.Index]*winmd.ImplMap),
+		FieldConstant:                 make(map[winmd.Index]*winmd.Constant),
+		TypeDefNativeTypedefAttribute: make(map[winmd.Index]*winmd.CustomAttribute),
+	}
+	// Scan some tables to create lookups for later.
+	for i := uint32(0); i < f.Tables.TypeDef.Len; i++ {
+		r, err := f.Tables.TypeDef.Record(winmd.Index(i))
+		if err != nil {
+			return nil, err
+		}
+		// TODO: Determine if disambiguating types in winmd files with multiple modules is necessary.
+		l.AllTypeDefs[r.Namespace.String()+"::"+r.Name.String()] = winmd.Index(i)
```

Comment:
It is a pity that `AllTypeDefs` uses a string as map key and that it needs to allocate a couple of times to generate the key, even more considering that  `r.Namespace` and `r.Name` could be uniquely identified using a 32-bits key into the Strings blob.

I'm wondering if the current String definition (`type String []byte`) is the right one, as a `[]byte` can't be used as map key. Using something like `type String struct { Index uint32, Data []byte }` would cover more use-cases. For example, `AllTypeDefs` could be defined as `map[struct{namespace uint32, name uint32}]winmd.Index`, which is much more efficient to store and lookup than a string.

Reply:
Done! I did this along with some changes that simplify how I'm resolving types by a lot.